test/filters/test_crop_hull.cpp

```diff
@@ -0,0 +1,586 @@
+      else {
+        mixedCloud->push_back(outsidePointGenerator());
+      }
+    }
+    testDataSuite.emplace_back(insideIndicesForInsideCloud, insideCloud);
+    testDataSuite.emplace_back(insideIndicesForOutsideCloud, outsideCloud);
+    testDataSuite.emplace_back(insideIndicesForMixedCloud, mixedCloud);
+    return testDataSuite;
+  }
+
+
+  class PCLCropHullTestFixtureBase : public ::testing::Test
+  {
+    public:
+      PCLCropHullTestFixtureBase() {
+        baseOffsetList.emplace_back(5, 1, 10);
+        baseOffsetList.emplace_back(1, 5, 10);
+        baseOffsetList.emplace_back(1, 10, 5);
+        baseOffsetList.emplace_back(10, 1, 5);
+        baseOffsetList.emplace_back(10, 5, 1);
+      }
+      static pcl::CropHull<pcl::PointXYZ> createDefaultCropHull (pcl::PointCloud<pcl::PointXYZ>::ConstPtr inputCloud) {
+        pcl::CropHull<pcl::PointXYZ> cropHullFilter(true);
+        pcl::ConvexHull<pcl::PointXYZ> convexHull;
+        convexHull.setDimension(3);
+        convexHull.setInputCloud(inputCloud);
+        pcl::PointCloud<pcl::PointXYZ>::Ptr hullCloudPtr(new pcl::PointCloud<pcl::PointXYZ>);
+        std::vector<pcl::Vertices> hullPolygons;
+        convexHull.reconstruct(*hullCloudPtr, hullPolygons);
+        cropHullFilter.setHullIndices(hullPolygons);
+        cropHullFilter.setHullCloud(hullCloudPtr);
+        return cropHullFilter;
+      }
+    protected:
+      pcl::PointCloud<pcl::PointXYZ> baseOffsetList;
+  };
+
+  class PCLCropHullTestFixture2d : public PCLCropHullTestFixtureBase
+  {
+    public:
+      PCLCropHullTestFixture2d()
+        : gen(12345u),
+          rd(0.0f, 1.0f),
+          baseInputCloud(new pcl::PointCloud<pcl::PointXYZ>)
+      {
+        baseInputCloud->emplace_back(0.0f, 0.0f, 0.0f);
+        baseInputCloud->emplace_back(0.0f, 1.0f, 0.0f);
+        baseInputCloud->emplace_back(1.0f, 0.0f, 0.0f);
+        baseInputCloud->emplace_back(1.0f, 1.0f, 0.0f);
+        baseInputCloud->emplace_back(0.0f, 0.0f, 0.1f);
+        baseInputCloud->emplace_back(0.0f, 1.0f, 0.1f);
+        baseInputCloud->emplace_back(1.0f, 0.0f, 0.1f);
+        baseInputCloud->emplace_back(1.0f, 1.0f, 0.1f);
+      }
+    protected:
+
+      void test(std::function<void(pcl::CropHull<pcl::PointXYZ>, TestData const &)> checker) {
+        pcl::PointCloud<pcl::PointXYZ>::Ptr inputCloud (new pcl::PointCloud<pcl::PointXYZ>);
+        for (pcl::PointXYZ const & baseOffset : baseOffsetList)
+        {
+          pcl::copyPointCloud(*baseInputCloud, *inputCloud);
+          for (pcl::PointXYZ & p : *inputCloud) {
+            p.getVector3fMap() += baseOffset.getVector3fMap();
+          }
+          auto insidePointGenerator = [this, &baseOffset] () {
+            pcl::PointXYZ p(rd(gen), rd(gen), 1.0);
+            p.getVector3fMap() += baseOffset.getVector3fMap();
+            return p;
+          };
+          auto outsidePointGenerator = [this, &baseOffset] () {
+            pcl::PointXYZ p(rd(gen) + 2., rd(gen) + 2., rd(gen) + 2.);
```

Comment:
Just confirming if the outside point gen needs the Z to be outside too. Else we can have the Z be 1.0 (to mirror L259)

Reply:
as it turned out, set Z value to constant in tests was bad idea.
after used random distribution for Z in L259, was found and fixed bug, which was described in issue #3960  (choice of dimensions in 2d case was based on input_cloud. now on hull_cloud)

Reply:
Nice. 😄 Maybe apply the modification in a separate PR though. This PR is spawning lots of additional PRs. Thanks for taking the time @DaniilsNikulin

Reply:
Keep the test here though